File: content/snippets/typescript/design_patterns/proxy.ts
class RealImage {
  constructor() {
    console.log('load image');
  }

  display(): void {
    console.log('display image');
  }
}

class ImageProxy {
  private real: RealImage | null = null;

  display(): void {
    if (!this.real) {
      this.real = new RealImage();
    }
    this.real.display();
  }
}

const proxy = new ImageProxy();
proxy.display();
proxy.display();

File: content/snippets/go/design_patterns/proxy.go
package main

import "fmt"

type Image interface {
	Display()
}

type RealImage struct{}

func NewRealImage() *RealImage {
	fmt.Println("load image")
	return &RealImage{}
}

func (RealImage) Display() {
	fmt.Println("display image")
}

type ImageProxy struct {
	real *RealImage
}

func (p *ImageProxy) Display() {
	if p.real == nil {
		p.real = NewRealImage()
	}
	p.real.Display()
}

func main() {
	proxy := &ImageProxy{}
	proxy.Display()
	proxy.Display()
}
